interpreter/ruby/ruby.go

```diff
@@ -289,21 +302,43 @@ func (r *rubyInstance) readPathObjRealPath(addr libpf.Address) (string, error) {
 	flags := r.rm.Ptr(addr)
 	switch flags & rubyTMask {
 	case rubyTString:
-		// nothing to do
+		return r.readRubyString(addr)
 	case rubyTArray:
-		var err error
-		addr, err = r.readRubyArrayDataPtr(addr)
-		if err != nil {
-			return "", err
+		vms := &r.r.vmStructs
+		arrData, e := r.readRubyArrayDataPtr(addr)
+		if e != nil {
+			return "", e
+		}
+
+		// Read contiguous pointer values into a buffer to be more efficient
+		dataBytes := make([]byte, 2 * vms.size_of_value)
+		if err := r.rm.Read(arrData, dataBytes); err != nil {
+			return "", fmt.Errorf("failed to read array data bytes: %v", err)
+		}
+
+		var relTag, absTag uint64
+		relVal := npsr.Ptr(dataBytes, 0)
+		absVal := npsr.Ptr(dataBytes, uint(vms.size_of_value))
+		if relVal != 0 {
+			relTag = uint64(r.rm.Ptr(relVal)) & uint64(rubyTMask)
+		}
+		if absVal != 0 {
+			absTag = uint64(r.rm.Ptr(absVal)) & uint64(rubyTMask)
+		}
```

Comment:
This is reading both `relVal` and `absVal` pointer unconditionally, but later `absVal` is preferred if it looks sane. Could this and the next block be combined so that only `absVal` is read and used if it looks ok, and `relVal` is read only if `absVal` is not sane. This would avoid one remote memory read.